controllers/topology/desired_state.go

```diff
@@ -173,14 +173,49 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 	}
 
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := contract.ControlPlane().Version().Set(controlPlane, s.Blueprint.Topology.Version); err != nil {
+	version, err := computeControlPlaneVersion(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute version of control plane")
+	}
+	if err := contract.ControlPlane().Version().Set(controlPlane, version); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion returns the desired control plane version.
+// If creating a new control plane, returns the version defined in tolopogy.
+// If a control plane already exists:
+// - If the control plane or any of the machine deployments are upgrading, returns the version from the current control plane.
+// - If not, returns the version defined in topology.
+func computeControlPlaneVersion(s *scope.Scope) (string, error) {
+	topologyVersion := s.Blueprint.Topology.Version
+	// creating a new control plane object, use version from topology
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		return topologyVersion, nil
+	}
+
+	version, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get the version from control plane spec")
+	}
+
+	// do not allow version downgrades
+	if err := isDowngrade(*version, topologyVersion); err != nil {
+		return *version, errors.Wrap(err, "failed version downgrade check")
+	}
+
+	cpUpgrading, err := contract.ControlPlane().Upgrading(s.Current.ControlPlane.Object)
+	if err != nil {
+		return *version, errors.Wrap(err, "failed to check if control plane is upgrading")
+	}
+	if cpUpgrading || s.Current.MachineDeployments.Upgrading() {
+		return *version, nil
+	}
+	return topologyVersion, nil
```

Comment:
Let's add comments also here, it will be verbose now, but it will facilitate maintenance in future;
For sake of readability, it might be also useful to break `if cpUpgrading || s.Current.MachineDeployments.Upgrading() {` into two if statements and provide explanation for the different use cases each parts of the if is trying to address